### PR TITLE
Added Question Answering Node

### PR DESCRIPTION
**Pull request by sshivaditya2019**
Added Question Node for Weaviate Neural Search. It can answer questions based on conversations stored as events.

### CODE REVIEW

File: packages/systems/src/systems/weaviate_events_schema.ts [hunk 1]
```diff
@@ -17,50 +23,62 @@ export default
...
+        "description": "Name of the channel through which event was created.",
+        "name": "channel"
       },
       {
         "dataType": [
           "string"
         ],
-        "description": "Contains the output of the text embedded in the text",
-        "name": "text"
+        "description": "Type of the channel thourgh which event was created.",
```
Comment (vincentfretin) [inline packages/systems/src/systems/weaviate_events_schema.ts:61]: thourgh -> through

File: packages/systems/src/systems/weaviate_events_schema.ts [hunk 2]
```diff
@@ -17,50 +23,62 @@ export default
...
           "string"
         ],
-        "description": "The date at which event was created",
+        "description": "The date at which the event was created",
         "name": "date"
+      },{
+        "dataType": [
+          "string"
+        ],
+        "description" : "The Observerver",
```
Comment (vincentfretin) [inline packages/systems/src/systems/weaviate_events_schema.ts:80]: Observerver -> Observer

File: packages/systems/src/systems/weaviateInit.ts [hunk 3]
```diff
@@ -161,4 +151,78 @@ export class weaviate_connection {
...
+                                        .explore()
+                                        .withNearText({
+                                          concepts: [data["concept"]],
+                                          distance: 0.6,
+                                          moveAwayFrom: {
+                                            concepts: [data["negative"]],
+                                            force: data["negative_distance"]
+                                          },
+                                          moveTo: {
+                                            concepts: [data["postive"]],
```
Comment (vincentfretin) [inline packages/systems/src/systems/weaviateInit.ts:212]: postive -> positive

File: packages/systems/src/systems/weaviateInit.ts [hunk 4]
```diff
@@ -161,4 +151,78 @@ export class weaviate_connection {
...
+    }
+    
+  }
+
+  static async semanticSearch(_data: SemanticSearch){
+    if (!weaviate_client) {
+        await initWeaviateClientEvent()
+    }
+    const data = { ..._data }
+    const validFields = ['concept', 'postive', 'negative', 'distance', 'postive_distance', 'negative_distance']
```
Comment (vincentfretin) [inline packages/systems/src/systems/weaviateInit.ts:196]: postive -> positive
postive_distance -> positive_distance

File: packages/core/types.ts [hunk 5]
```diff
@@ -52,6 +52,19 @@ export type Event = {
   date?: string
 }
 
+export type SemanticSearch = {
+  concept?: string
+  postive?: string
```
Comment (vincentfretin) [inline packages/core/types.ts:57]: postive -> positive

File: packages/core/types.ts [hunk 6]
```diff
@@ -52,6 +52,19 @@ export type Event = {
...
+  negative?: string
+  distance?: number
+  postive_distance?: number
```
Comment (vincentfretin) [inline packages/core/types.ts:60]: postive_distance -> positive_distance

**Comment by lalalune** [comment]
@sshivaditya had a chance to review

- Storing events works well.

- Unable to recall events using the event recall node

- QA works well

Two things are needed -- a way to delete events and the ability to recall and filter by the entities.

The problem is that when the agent speakers, the sender and observer are both agent. To keep track of which conversation we are in, we need to key it by the entities -- speaker and agent at the very least, but there could be 3 speakers and an agent. We'd want the last conversation where the entities include both speaker and agent, whether other people came or left.

We are doing this conversation in a 3D world where other users and agents might walk up, so tracking just 2 speakers in the convo doesn't make sense.

We also have multiple agents, so when we ask the question, we need to make sure it's keyed to only recall the information the current agent knows.

**Comment by sshivaditya2019** [comment]
> Unable to recall events using the event recall node

Recall Node was returning the events in Array format. Fixed by converting into JSON and returning, the same.

> Two things are needed -- a way to delete events and the ability to recall and filter by the entities.

Added a Delete Node, needs all values same as the store Node. These values are used to get the ID of all matching events, which are then deleted.

Now, the event recall will filter events based on observer and sender, only those conversations where the sender, observer and agent were active will be returned.

> We also have multiple agents, so when we ask the question, we need to make sure it's keyed to only recall the information the current agent knows

Added an agent id based filter for the question answer node.

**Pull request merged**